src/main/scala/viper/gobra/translator/encodings/preds/PredEncoding.scala

```diff
@@ -0,0 +1,175 @@
+import viper.gobra.theory.Addressability.{Exclusive, Shared}
+import viper.gobra.translator.interfaces.{Collector, Context}
+import viper.gobra.translator.util.ViperWriter.CodeWriter
+import viper.silver.{ast => vpr}
+import viper.silver.verifier.{errors => vprerr}
+
+class PredEncoding extends LeafTypeEncoding {
+
+  import viper.gobra.translator.util.ViperWriter.CodeLevel._
+  import viper.gobra.translator.util.TypePatterns._
+
+  private val defunc: DefuncComponent = new DefuncComponentImpl
+
+  override def finalize(col: Collector): Unit = {
+    defunc.finalize(col)
+  }
+
+  /**
+    * Translates a type into a Viper type.
+    */
+  override def typ(ctx: Context): in.Type ==> vpr.Type = {
+    case ctx.Pred(ts) / m =>
+      m match {
+        case Exclusive => defunc.typ(ts)(ctx)
+        case Shared    => vpr.Ref: vpr.Type
+      }
+  }
+
+  /**
+    * Encodes expressions as values that do not occupy some identifiable location in memory.
+    *
+    * To avoid conflicts with other encodings, a leaf encoding for type T should be defined at:
+    * (1) exclusive operations on T, which includes literals and default values
+    *
+    * [Q{d1, ..., dk}: pred(S)] -> make_S_ID([d1], ..., [dk]) where ID is the ID for the pattern used in Q{d1, ..., dk}
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.expr(ctx)) {
+      case n@ in.PredicateConstructor(p, pTs, args) :: ctx.Pred(_) / Exclusive =>
+        val (pos, info, errT) = n.vprMeta
+        for {
+          vArgs <- sequence(args map (a => option(a map goE)))
+        } yield defunc.construct(p, pTs.args, vArgs)(pos, info, errT)(ctx)
+    }
+  }
+
+  /**
+    * Encodes assertions.
+    *
+    * Constraints:
+    * - in.Access with in.PredicateAccess has to encode to vpr.PredicateAccessPredicate.
+    *
+    * [acc(p(e1, ..., en))] -> eval_S([p], [e1], ..., [en]) where p: pred(S)
+    */
+  override def assertion(ctx: Context): in.Assertion ==> CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.assertion(ctx)) {
+      case n@ in.Access(in.Accessible.PredExpr(in.PredExprInstance(p :: ctx.Pred(ts), args)), perm) =>
+        val (pos, info, errT) = n.vprMeta
+        for {
+          vArgs <- sequence(args map goE)
+          vBase <- goE(p)
+          vPerm <- goE(perm)
+        } yield vpr.PredicateAccessPredicate(defunc.instance(vBase, ts, vArgs)(pos, info, errT)(ctx), vPerm)(pos, info, errT) : vpr.Exp
+    }
+  }
+
+  /**
+    * Encodes statements.
+    * This includes make-statements.
+    *
+    * The default implements:
+    * [v: *T = make(lit)] -> var z (*T)°; inhale Footprint[*z] && [*z == lit]; [v = z]
+    *
+    * [unfold acc(Q{d1, ..., dk}(e1, ..., en), p)] ->
+    *   exhale acc(eval_S([Q{d1, ..., dk}], [e1], ..., [en]), [p])
+    *   inhale acc(Q(a1, ..., ak), [p])
```

Comment:
Is this exhale/inhale sequence the reason why the `eval_S` predicate can be abstract? Also, is this a consequence of the restriction of knowing the base when folding/unfolding?

Reply:
I would state this a bit differently. There is a soundness issue when you do not know the base. But we cannot write a `eval_S` that way in Viper, so we make it abstract. Because it is abstract, we cannot fold and unfold, but have to exhale and inhale instead.

Reply:
Does this mean that the soundness issue prevents writing eval_S‘s body as a case split over the IDs? I.e. having an unknown case is not enough